src/parsers.rs

```diff
@@ -1,13 +1,335 @@
 
+use crate::{
+    options::{DisplayCalendar, DisplayOffset, DisplayTimeZone},
+    Sign, TemporalError, TemporalResult, TemporalUnwrap,
+};
 use alloc::format;
-
-use crate::{TemporalError, TemporalResult, TemporalUnwrap};
-
 use ixdtf::parsers::{
     records::{Annotation, DateRecord, IxdtfParseRecord, TimeRecord, UtcOffsetRecordOrZ},
     IxdtfParser,
 };
+use writeable::{impl_display_with_writeable, LengthHint, Writeable};
+
+// TODO: Move `Writeable` functionality to `ixdtf` crate
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Precision {
+    Auto,
+    Minute,
+    Digit(u8),
+}
+
+pub struct FormattableTime {
+    pub hour: u8,
+    pub minute: u8,
+    pub second: u8,
+    pub nanosecond: u32,
+    pub precision: Precision,
+    pub include_sep: bool,
+}
+
+impl Writeable for FormattableTime {
+    fn write_to<W: core::fmt::Write + ?Sized>(&self, sink: &mut W) -> core::fmt::Result {
+        write_padded_u8(self.hour, sink)?;
+        if self.include_sep {
+            sink.write_char(':')?;
+        }
+        write_padded_u8(self.minute, sink)?;
+        if self.precision == Precision::Minute {
+            return Ok(());
+        }
+        if self.include_sep {
+            sink.write_char(':')?;
+        }
+        write_padded_u8(self.second, sink)?;
+        if self.nanosecond == 0 {
+            return Ok(());
+        }
+        sink.write_char('.')?;
+        write_nanosecond(self.nanosecond, self.precision, sink)
+    }
+
+    fn writeable_length_hint(&self) -> LengthHint {
+        let sep = self.include_sep as usize;
+        if self.precision == Precision::Minute {
+            return LengthHint::exact(4 + sep);
+        }
+        let time_base = 6 + (sep * 2);
+        if self.nanosecond == 0 {
+            return LengthHint::exact(time_base);
+        }
+        if let Precision::Digit(d) = self.precision {
+            return LengthHint::exact(time_base + 1 + d as usize);
+        }
+        LengthHint::between(time_base + 2, time_base + 10)
+    }
+}
+
+pub struct FormattableUtcOffset {
+    pub show: DisplayOffset,
+    pub offset: UtcOffset,
+}
+
+pub enum UtcOffset {
+    Z,
+    Offset(FormattableOffset),
+}
+
+impl Writeable for FormattableUtcOffset {
+    fn write_to<W: core::fmt::Write + ?Sized>(&self, sink: &mut W) -> core::fmt::Result {
+        if self.show == DisplayOffset::Never {
+            return Ok(());
+        }
+        match &self.offset {
+            UtcOffset::Z => sink.write_char('Z'),
+            UtcOffset::Offset(offset) => offset.write_to(sink),
+        }
+    }
+
+    fn writeable_length_hint(&self) -> LengthHint {
+        match &self.offset {
+            UtcOffset::Z => LengthHint::exact(1),
+            UtcOffset::Offset(o) => o.writeable_length_hint(),
+        }
+    }
+}
+
+fn write_padded_u8<W: core::fmt::Write + ?Sized>(num: u8, sink: &mut W) -> core::fmt::Result {
+    if num < 10 {
+        sink.write_char('0')?;
+    }
+    num.write_to(sink)
+}
```

Comment:
Suggestion: I like how you use `impl Writeable for u8` here. I think you can do that with the u32 as well.

Something like:

```rust
let mut max = 1e9;
while max > nanoseconds {
    sink.write_char('0')?;
    max /= 10;
}
nanoseconds.write_to(sink)?;
```

And then you don't need any unsafe code.

Reply:
Does that fix output the nanoseconds according to what would be expected? If I'm reading that write, it would left pad nanoseconds to always be 9 digits, but the concern with this is to trim trailing zeros.

nanoseconds = 123_000_000, should only print .123 or the desired precision, i.e. precision = 5 would be .12300, but this would print 123000000

The idea for the use of unsafe in the function was to try and optimize out the `div` instruction from the below, which I think would be the equivalent to the above.

```
   let mut divisor = 1_000_000_000;
    while divisor >= 1 && nanoseconds != 0 {
        (nanoseconds / divisor).write_to(sink)?;
        nanoseconds %= divisor;
        divisor /= 10;
    }
```

Reply:
Actually, nevermind. We can just write the digits manually as `u8`s